### PR TITLE
Add multiple improvements and fixes

### PR DESCRIPTION
**Pull request by p4checo**
### Motivation

Found an issue on iOS Release builds, where the wrong connection status is reported after returning from background on device sleep, and doesn't recover unless one manually changes network settings.

While I initially thought the problem would be on `Reachability`, I came to the conclusion that when `SCNetworkReachability`) is configured with a hostname and callback monitoring is enabled, and the device is locked or auto-locks (and the app goes to background), sometimes the reachability callback will report empty flags (Not Reachable), and won't recover afterwards.

My suspicion is that this is some problem with the DNS background service interfering with the `SCNetworkReachability` (revealed by the `dnssd_clientstub`and `nw_resolver_create_dns_service_locked` logs that show up when the problem occurs), because this issue only seems to manifest itself when using a hostname to configure the Reachability.

To mitigate this, the `SCNetworkReachability` can simply be instantiated with an address (0.0.0.0), which seems to fix the issue given it doesn't depend on DNS.

For more information please see the reported Radar [rdar://40291350](https://openradar.appspot.com/40291350).

While tracking down the issue, I ended up making some changes and improvements along the way 😇 

### Changes

- Updated `flags` to be optional and a stored property which is updated when the callback fires and not fetched every time it's accessed (computed). Besides being inefficient, getting the flags (via `SCNetworkReachabilityGetFlags`) on every access most likely caused the issue mentioned above, because after fetching some correct values (`2` - reachable), it would then return one or more incorrect values (`0` - none) which would trigger the wrong state to be made available.

- The `reachabilityChanged` is now invoked on the `flags` `didSet`.

- Made the callback a closure instead of a static function, which now also updates the `flags` property instead of triggering `reachabilityChanged` directly.

- Moved the `isXFlagSet` getters to a `SCNetworkReachabilityFlags` extension.

- Made the initial check synchronous and throwing a new `ReachabilityError.UnableToGetInitialFlags` error when obtaining the initial flags fails.

- Made the internal queue more configurable, by allowing setting its QoS and target queue on `init`.

- Did some cleanup, like removing logs and unnecessary vars, simplifying legacy getters, avoiding possible retain cycles.

- Fixed the invalid hostname unit test, and did some cleanup.

- Fixed .podspec's `source_files` path.

### CODE REVIEW

**Comment by hamchapman** [comment]
This looks great - I hope to be able to give it a proper review soon 👍

**Comment by p4checo** [comment]
That's nice to hear! Looking forward to it 😉 

Unfortunately, I've made some further testing and the issue I mentioned above still persists even with these changes. It sucks, but I think the changes still make sense and bring some benefit.

I'm still investigating the issue and will report back if I find anything new.

So far, the best clue I have is that the issue is probably caused by the iOS's DNS service becoming unresponsive/crashing/being restarted, because I saw these logs one of the times it happened:

```
2018-05-11 17:21:46.817873+0100 <MyApp>[1134:272990] Task <410096F4-D81F-4A8A-92C8-E6991B48D625>.<5> finished with error - code: -1001
2018-05-11 17:21:47.355143+0100 <MyApp>[1134:269088] dnssd_clientstub read_all(11) DEFUNCT
2018-05-11 17:21:47.355412+0100 <MyApp>[1134:269088] [] nw_resolver_create_dns_service_locked DNSServiceCreateDelegateConnection failed: ServiceNotRunning(-65563)
2018-05-11 17:33:51.517379+0100 <MyApp>[1134:269088] dnssd_clientstub read_all(6) DEFUNCT
2018-05-11 17:33:51.517466+0100 <MyApp>[1134:269088] [] nw_resolver_create_dns_service_locked DNSServiceCreateDelegateConnection failed: ServiceNotRunning(-65563)
2018-05-11 17:33:51.616458+0100 <MyApp>[1134:269127] dnssd_clientstub read_all(25) DEFUNCT
2018-05-11 17:33:51.616516+0100 <MyApp>[1134:269127] [] nw_resolver_create_dns_service_locked DNSServiceCreateDelegateConnection failed: ServiceNotRunning(-65563)
2018-05-11 17:33:51.624980+0100 <MyApp>[1134:272993] dnssd_clientstub read_all(26) DEFUNCT
2018-05-11 17:33:51.632089+0100 <MyApp>[1134:272993] dnssd_clientstub write_all(7) DEFUNCT
2018-05-11 17:33:51.632437+0100 <MyApp>[1134:272993] [] nw_resolver_create_dns_service_locked DNSServiceGetAddrInfo failed: ServiceNotRunning(-65563)
received flags: SCNetworkReachabilityFlags(rawValue: 0), decription: -- -------
updated flags to: Optional(__C.SCNetworkReachabilityFlags(rawValue: 0)), decription: Optional("-- -------")
2018-05-11 17:33:51.641402+0100 <MyApp>[1134:272999] dnssd_clientstub read_all(23) DEFUNCT
2018-05-11 17:33:51.641758+0100 <MyApp>[1134:272999] dnssd_clientstub read_all(24) DEFUNCT
```

File: Sources/Reachability.swift [hunk 1]
```diff
@@ -207,117 +194,115 @@ public extension Reachability {
...
-        let W = isRunningOnDevice ? (isOnWWANFlagSet ? "W" : "-") : "X"
-        let R = isReachableFlagSet ? "R" : "-"
-        let c = isConnectionRequiredFlagSet ? "c" : "-"
-        let t = isTransientConnectionFlagSet ? "t" : "-"
-        let i = isInterventionRequiredFlagSet ? "i" : "-"
-        let C = isConnectionOnTrafficFlagSet ? "C" : "-"
-        let D = isConnectionOnDemandFlagSet ? "D" : "-"
-        let l = isLocalAddressFlagSet ? "l" : "-"
-        let d = isDirectFlagSet ? "d" : "-"
+        guard let flags = flags else { return "unavaiable flags" }
```
Comment (LukeDurrant) [inline Sources/Reachability.swift]: `unavaiable` -> `unavailable`

**Comment by ashleymills** [comment]
@p4checo Thanks for this fantastic PR

Cheers
Ash

**Comment by p4checo** [comment]
Cheers! 🍻 
Thanks for the cool project! 🙏

**Pull request merged**